File: src/numeric/max/max.pas
{
    Get the maximum value of a sequence of values
}

program max;
{$mode objfpc} // generics is a freepascal feature; we break other pascal possibly here
    uses SysUtils;

    (* Get the maximum value of a sequence of values of type T *)
    generic function max<T>(values : array of T) : T;
        var
            current : T;
            i : integer;
    begin
        current := values[0];
        for i := Low(values) to High(values) do begin
            if values[i] > current then
                current := values[i];
        end;
        Result := current;
    end;

    var values : array of integer;
    var i, maximum : integer;
begin
    if ParamCount > 0 then begin
        SetLength(values, ParamCount);
        for i := 1 to ParamCount do begin
            values[i - 1] := StrToInt(ParamStr(i));
        end;
    end
    else begin
        SetLength(values, 2);
        values[0] := 15;
        values[1] := 10;
    end;

    maximum := specialize max<integer>(values);

    Write('values: [', values[0]);
    for i := 1 to High(values) do begin
        Write(', ', values[i]);
    end;
    WriteLn(']');
    WriteLn('max: ', maximum);
end.

File: src/numeric/max/max.nim
#[
    Get the maximum value of some sequence of values
]#
import os
import std/strutils
import std/strformat

# Get the maximum value of a sequence
proc max[T](values: var seq[T]): T =
    var current = values[0]
    for value in values:
        if value > current:
            current = value
    return current

var values = newSeq[int]()
if paramCount() > 0:
    for i in 1..paramCount():
        values.add(parseInt(paramStr(i)))
else:
    values.add(15)
    values.add(10)

let maximum = max(values)

echo &"values: {values}"
echo &"max: {maximum}"
